neuro_evolution_ctrnn/brains/i_layer_based_brain.py

```diff
@@ -0,0 +1,190 @@
+                        individual_index:
+                        individual_index + hidden_structure[layer]
+                    ]
+                ))
+                individual_index += hidden_structure[layer]
+            else:
+                self.hidden.append(np.zeros((hidden_structure[layer])))
+            self.layer_output.append(np.zeros((hidden_structure[layer])))
+        # for end
+
+        # Matrix for transforming output of last layer into output neurons
+        self.weight_ho = np.array(
+            individual[
+                individual_index:
+                individual_index + hidden_structure[len(hidden_structure) - 1] * output_size
+            ]
+        ).reshape((output_size, hidden_structure[len(hidden_structure) - 1]))
+        individual_index += hidden_structure[len(hidden_structure) - 1] * output_size
+
+        # Has all values been used and therefore does get_individual_size() provide the right number?
+        assert individual_index == len(individual)
+
+    @staticmethod
+    @abc.abstractmethod
+    def get_number_gates():
+        # How many Gates are used in the specific network?
+        # Haw many matrices are needed for each layer to calculate the next state and output value
+        pass
+
+    @classmethod
+    def get_individual_size(cls, config: ILayerBasedBrainCfg, input_space: Space, output_space: Space):
+
+        number_gates = cls.get_number_gates()
+        input_size = cls._size_from_space(input_space)
+        output_size = cls._size_from_space(output_space)
+        hidden_struc = config.hidden_structure
+
+        individual_size = 0
+
+        for layer in range(len(hidden_struc)):
+            # Matrices for weighted input values
+            if layer == 0:  # The first Layer don't has an output from the previous layer, but the input values
+                individual_size += number_gates * input_size * hidden_struc[0]
+            else:
+                individual_size += number_gates * hidden_struc[layer] * hidden_struc[layer - 1]
+
+            # Matrices for weighted state values
+            if config.diagonal_hidden_to_hidden:
+                individual_size += number_gates * hidden_struc[layer]
+            else:
+                individual_size += number_gates * hidden_struc[layer] * hidden_struc[layer]
+
+            # initialize biases
+            if config.use_bias:
+                individual_size += hidden_struc[layer] * number_gates
+
+            # Hidden values
+            if config.optimize_initial_hidden_values:
+                individual_size += hidden_struc[layer]
+        # for end
+
+        # Matrix for transforming output of last layer into output neurons
+        individual_size += hidden_struc[len(hidden_struc) - 1] * output_size
+        return individual_size
+
+    def brainstep(self, ob: np.ndarray):
+
+        layer_input = ob
+        # iterate for all given layers
+        for layer in range(len(self.hidden)):
+            if layer == 0:
```

Comment:
ggf. vor die schleife schreiben. 

Falls Ja: beachte, dass in Z.36 die Struktur vom Code so ähnlich zu dieser stelle sein sollte wie möglich. Also wenn du es hier außerhalb der schleife machst, dann oben auch (und umgekehrt)